File: 03-classic-algorithms/sorting/insertsort/insert_sort.go
/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */
package main

import "fmt"

/**
 * 插入排序标准版
 */
func insertSort1(arr []int) []int {
  var j int
  var current int
  var l = len(arr)

  for i := 0; i < l; i++ {
    // 当前项为第1项
    current = arr[i]
    // j是已排序的末尾位置
    j = i - 1

    // 内循环用来遍历已排序区间
    // 将该项自右往左与已排序项逐个对比，当遇到比自己大的项时，则将该位置逐项右移
    // 直到遇到小于等于自己的项则停止移动，表示插入成功
    fmt.Println("i=", i, " j=", j, "current=", current, "arr[i]=", arr[i], "arr[]=", arr)
    for j >= 0 && current < arr[j] {
      // 逐项右移，以便空出插入项
      arr[j+1] = arr[j]
      j--
    }

    // 没有比自己大的项了，则插入到当前位置
    arr[j+1] = current
  }
  return arr
}

/**
 * 插入排序降序版
 */
func insertSort2(arr []int) []int {
  for i := 0; i < len(arr); i++ {
    var current = arr[i]
    var j = i - 1
    fmt.Println("i=", i, " j=", j, "current=", current, "arr[i]=", arr[i], "arr[j + 1]=", arr[j+1], "arr[]=", arr)
    for ; j >= 0; j-- {
      // 当前项比已排序的内容要大，则逐个右移，空出位置
      if current > arr[j] {
        arr[j+1] = arr[j]
      } else {
        // 当小于已排序内容，则跳出循环
        break
      }
    }
    arr[j+1] = current
  }
  return arr
}

/**
 * 插入排序直接插入法，利用copy函数
 */
func insertSort3(arr []int) []int {
  for i, current := range arr {
    var j = i - 1
    isMin := true
    fmt.Println("i=", i, " j=", j, "current=", current, "arr[i]=", arr[i], "arr[j + 1]=", arr[j+1], "arr[]=", arr)
    for j >= 0 {
      // 当前项比已排序的内容要大，则插入其后
      if current > arr[j] {
        // 将j至i全部项整体右移1位
        copy(arr[j+1:i+1], arr[j:i])
        arr[j+1] = current
        isMin = false
        break
      }
      j -= 1
    }
    // 如果当前项最小则插入到已排序的最前面
    if isMin {
      copy(arr[1:i+1], arr[0:i])
      arr[0] = current
    }
  }
  return arr
}

func main() {
  fmt.Println("insert sort1:")
  data := [...]int{7, 11, 9, 10, 12, 13, 8}
  data2 := [...]int{3, 4, 15, -2, -2, -1, 2, 6, 3, -7, 10}
  fmt.Println(insertSort1(data[:]))
  fmt.Println(insertSort1(data2[:]))

  fmt.Println("insert sort2:")
  data = [...]int{7, 11, 9, 10, 12, 13, 8}
  data2 = [...]int{3, 4, 15, -2, -2, -1, 2, 6, 3, -7, 10}
  fmt.Println(insertSort2(data[:]))
  fmt.Println(insertSort2(data2[:]))

  fmt.Println("insert sort3:")
  data = [...]int{7, 11, 9, 10, 12, 13, 8}
  data2 = [...]int{3, 4, 15, -2, -2, -1, 2, 6, 3, -7, 10}
  fmt.Println(insertSort3(data[:]))
  fmt.Println(insertSort3(data2[:]))
}

/* execute test
jarry@jarrys-MacBook-Pro insertsort % go run insert_sort.go
insert sort1:
i= 0  j= -1 current= 7 arr[i]= 7 arr[]= [7 11 9 10 12 13 8]
i= 1  j= 0 current= 11 arr[i]= 11 arr[]= [7 11 9 10 12 13 8]
i= 2  j= 1 current= 9 arr[i]= 9 arr[]= [7 11 9 10 12 13 8]
i= 3  j= 2 current= 10 arr[i]= 10 arr[]= [7 9 11 10 12 13 8]
i= 4  j= 3 current= 12 arr[i]= 12 arr[]= [7 9 10 11 12 13 8]
i= 5  j= 4 current= 13 arr[i]= 13 arr[]= [7 9 10 11 12 13 8]
i= 6  j= 5 current= 8 arr[i]= 8 arr[]= [7 9 10 11 12 13 8]
[7 8 9 10 11 12 13]
i= 0  j= -1 current= 3 arr[i]= 3 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 1  j= 0 current= 4 arr[i]= 4 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 2  j= 1 current= 15 arr[i]= 15 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 3  j= 2 current= -2 arr[i]= -2 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 4  j= 3 current= -2 arr[i]= -2 arr[]= [-2 3 4 15 -2 -1 2 6 3 -7 10]
i= 5  j= 4 current= -1 arr[i]= -1 arr[]= [-2 -2 3 4 15 -1 2 6 3 -7 10]
i= 6  j= 5 current= 2 arr[i]= 2 arr[]= [-2 -2 -1 3 4 15 2 6 3 -7 10]
i= 7  j= 6 current= 6 arr[i]= 6 arr[]= [-2 -2 -1 2 3 4 15 6 3 -7 10]
i= 8  j= 7 current= 3 arr[i]= 3 arr[]= [-2 -2 -1 2 3 4 6 15 3 -7 10]
i= 9  j= 8 current= -7 arr[i]= -7 arr[]= [-2 -2 -1 2 3 3 4 6 15 -7 10]
i= 10  j= 9 current= 10 arr[i]= 10 arr[]= [-7 -2 -2 -1 2 3 3 4 6 15 10]
[-7 -2 -2 -1 2 3 3 4 6 10 15]
insert sort2:
i= 0  j= -1 current= 7 arr[i]= 7 arr[j + 1]= 7 arr[]= [7 11 9 10 12 13 8]
i= 1  j= 0 current= 11 arr[i]= 11 arr[j + 1]= 11 arr[]= [7 11 9 10 12 13 8]
i= 2  j= 1 current= 9 arr[i]= 9 arr[j + 1]= 9 arr[]= [11 7 9 10 12 13 8]
i= 3  j= 2 current= 10 arr[i]= 10 arr[j + 1]= 10 arr[]= [11 9 7 10 12 13 8]
i= 4  j= 3 current= 12 arr[i]= 12 arr[j + 1]= 12 arr[]= [11 10 9 7 12 13 8]
i= 5  j= 4 current= 13 arr[i]= 13 arr[j + 1]= 13 arr[]= [12 11 10 9 7 13 8]
i= 6  j= 5 current= 8 arr[i]= 8 arr[j + 1]= 8 arr[]= [13 12 11 10 9 7 8]
[13 12 11 10 9 8 7]
i= 0  j= -1 current= 3 arr[i]= 3 arr[j + 1]= 3 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 1  j= 0 current= 4 arr[i]= 4 arr[j + 1]= 4 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 2  j= 1 current= 15 arr[i]= 15 arr[j + 1]= 15 arr[]= [4 3 15 -2 -2 -1 2 6 3 -7 10]
i= 3  j= 2 current= -2 arr[i]= -2 arr[j + 1]= -2 arr[]= [15 4 3 -2 -2 -1 2 6 3 -7 10]
i= 4  j= 3 current= -2 arr[i]= -2 arr[j + 1]= -2 arr[]= [15 4 3 -2 -2 -1 2 6 3 -7 10]
i= 5  j= 4 current= -1 arr[i]= -1 arr[j + 1]= -1 arr[]= [15 4 3 -2 -2 -1 2 6 3 -7 10]
i= 6  j= 5 current= 2 arr[i]= 2 arr[j + 1]= 2 arr[]= [15 4 3 -1 -2 -2 2 6 3 -7 10]
i= 7  j= 6 current= 6 arr[i]= 6 arr[j + 1]= 6 arr[]= [15 4 3 2 -1 -2 -2 6 3 -7 10]
i= 8  j= 7 current= 3 arr[i]= 3 arr[j + 1]= 3 arr[]= [15 6 4 3 2 -1 -2 -2 3 -7 10]
i= 9  j= 8 current= -7 arr[i]= -7 arr[j + 1]= -7 arr[]= [15 6 4 3 3 2 -1 -2 -2 -7 10]
i= 10  j= 9 current= 10 arr[i]= 10 arr[j + 1]= 10 arr[]= [15 6 4 3 3 2 -1 -2 -2 -7 10]
[15 10 6 4 3 3 2 -1 -2 -2 -7]
insert sort3:
i= 0  j= -1 current= 7 arr[i]= 7 arr[j + 1]= 7 arr[]= [7 11 9 10 12 13 8]
i= 1  j= 0 current= 11 arr[i]= 11 arr[j + 1]= 11 arr[]= [7 11 9 10 12 13 8]
i= 2  j= 1 current= 9 arr[i]= 9 arr[j + 1]= 9 arr[]= [7 11 9 10 12 13 8]
i= 3  j= 2 current= 10 arr[i]= 10 arr[j + 1]= 10 arr[]= [7 9 11 10 12 13 8]
i= 4  j= 3 current= 12 arr[i]= 12 arr[j + 1]= 12 arr[]= [7 9 10 11 12 13 8]
i= 5  j= 4 current= 13 arr[i]= 13 arr[j + 1]= 13 arr[]= [7 9 10 11 12 13 8]
i= 6  j= 5 current= 8 arr[i]= 8 arr[j + 1]= 8 arr[]= [7 9 10 11 12 13 8]
[7 8 9 10 11 12 13]
i= 0  j= -1 current= 3 arr[i]= 3 arr[j + 1]= 3 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 1  j= 0 current= 4 arr[i]= 4 arr[j + 1]= 4 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 2  j= 1 current= 15 arr[i]= 15 arr[j + 1]= 15 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 3  j= 2 current= -2 arr[i]= -2 arr[j + 1]= -2 arr[]= [3 4 15 -2 -2 -1 2 6 3 -7 10]
i= 4  j= 3 current= -2 arr[i]= -2 arr[j + 1]= -2 arr[]= [-2 3 4 15 -2 -1 2 6 3 -7 10]
i= 5  j= 4 current= -1 arr[i]= -1 arr[j + 1]= -1 arr[]= [-2 -2 3 4 15 -1 2 6 3 -7 10]
i= 6  j= 5 current= 2 arr[i]= 2 arr[j + 1]= 2 arr[]= [-2 -2 -1 3 4 15 2 6 3 -7 10]
i= 7  j= 6 current= 6 arr[i]= 6 arr[j + 1]= 6 arr[]= [-2 -2 -1 2 3 4 15 6 3 -7 10]
i= 8  j= 7 current= 3 arr[i]= 3 arr[j + 1]= 3 arr[]= [-2 -2 -1 2 3 4 6 15 3 -7 10]
i= 9  j= 8 current= -7 arr[i]= -7 arr[j + 1]= -7 arr[]= [-2 -2 -1 2 3 3 4 6 15 -7 10]
i= 10  j= 9 current= 10 arr[i]= 10 arr[j + 1]= 10 arr[]= [-7 -2 -2 -1 2 3 3 4 6 15 10]
[-7 -2 -2 -1 2 3 3 4 6 10 15]
*/

File: 03-classic-algorithms/sorting/insertsort/insert_sort.py
#!/usr/bin/env python
#-*- encoding: UTF-8-*-
"""
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
"""
import time

'''
# 插入排序标准通用版
'''
def insert_sort1(arr):
    l = len(arr)
    for i in range(l):
        j = i - 1
        current = arr[i]
        # 将该项自右往左与已排序项逐个对比，当遇到比自己大的项时，将该位置右移
        # 直到遇到小于等于自己的项则停止移动，表示插入成功
        print('i=' + str(i), ' j=' + str(j), 'current=' + str(current),
              'arr[i]=' + str(arr[i]), 'arr[j]=' + str(arr[j]), 'arr[]=', arr)
        while (j >= 0 and current < arr[j]):
            # 逐个位移
            arr[j + 1] = arr[j]
            j -= 1
        # 交换为当前项
        arr[j + 1] = current
    return arr

'''
# 插入排序利用list insert函数插入选择项
'''
def insert_sort2(arr):
    l = len(arr)
    is_min = False
    for i in range(l):
        j = i - 1
        current = arr[i]
        print('current:', current)
        is_min = True
        print('i=' + str(i), ' j=' + str(j), 'current=' + str(current),
              'arr[i]=' + str(arr[i]), 'arr[j]=' + str(arr[j]), 'arr[]=', arr)
        while j >= 0:
            if (current > arr[j]):
                # 利用insert插入
                arr.insert(j + 1, current)
                is_min = False
                break
            j -= 1
        if (is_min):
            arr.insert(0, current)
        arr.pop(i)
    return arr

'''
# 插入排序利用list insert函数插入最大项
'''
def insert_sort3(arr):
    l = len(arr)
    new_arr = [arr[0]]
    for i in range(1, l):
        isMax = True
        for j in range(len(new_arr)):
            print('i=' + str(i), ' j=' + str(j), 'arr[i]=' + str(arr[i]),
                  'new_arr[j]=' + str(new_arr[j]), 'new_arr[]=', new_arr)
            if arr[i] < new_arr[j]:
                # 新建数组来插入到某个位置中
                new_arr.insert(j, arr[i])
                isMax = False
                break
        if (isMax == True):
            new_arr.append(arr[i])
    print('new_arr:', new_arr)
    return new_arr


'''
# 插入排序新建数组版倒序遍历
'''
def insert_sort4(arr):
    l = len(arr)
    new_arr = [arr[0]]
    for i in range(1, l):
        is_min = True
        new_len = len(new_arr)
        while (new_len > 0):
            new_len -= 1
            print('i=' + str(i), ' new_len=' + str(new_len),
                  'arr[i]=' + str(arr[i]),
                  'new_arr[new_len]=' + str(new_arr[new_len]), 'new_arr[]=', new_arr)
            if arr[i] > new_arr[new_len]:
                new_arr.insert(new_len + 1, arr[i])
                is_min = False
                break
        if is_min == True:
            new_arr.insert(0, arr[i])
    print('new_arr:', new_arr)
    return new_arr


# test
arr1 = [7, 11, 9, 10, 12, 13, 8]
start_time = time.time()
print(insert_sort1(arr1))
print("time:" + str((time.time() - start_time) * 1000) + " ms")

arr2 = [7, 11, 9, 10, 12, 13, 8]
start_time = time.time()
print(insert_sort2(arr2))
print("time:" + str((time.time() - start_time) * 1000) + " ms")

arr3 = [7, 11, 9, 10, 12, 13, 8]
start_time = time.time()
print(insert_sort3(arr3))
print("time:" + str((time.time() - start_time) * 1000) + " ms")

arr4 = [7, 11, 9, 10, 12, 13, 8]
start_time = time.time()
print(insert_sort4(arr4))
print("time:" + str((time.time() - start_time) * 1000) + " ms")
"""
jarrys-MacBook-Pro:insert_sort jarry$ python insert_sort.py
('i=0', ' j=-1', 'current=7', 'arr[i]=7', 'arr[j]=8', 'arr[]=', [7, 11, 9, 10, 12, 13, 8])
('i=1', ' j=0', 'current=11', 'arr[i]=11', 'arr[j]=7', 'arr[]=', [7, 11, 9, 10, 12, 13, 8])
('i=2', ' j=1', 'current=9', 'arr[i]=9', 'arr[j]=11', 'arr[]=', [7, 11, 9, 10, 12, 13, 8])
('i=3', ' j=2', 'current=10', 'arr[i]=10', 'arr[j]=11', 'arr[]=', [7, 9, 11, 10, 12, 13, 8])
('i=4', ' j=3', 'current=12', 'arr[i]=12', 'arr[j]=11', 'arr[]=', [7, 9, 10, 11, 12, 13, 8])
('i=5', ' j=4', 'current=13', 'arr[i]=13', 'arr[j]=12', 'arr[]=', [7, 9, 10, 11, 12, 13, 8])
('i=6', ' j=5', 'current=8', 'arr[i]=8', 'arr[j]=13', 'arr[]=', [7, 9, 10, 11, 12, 13, 8])
[7, 8, 9, 10, 11, 12, 13]
time:0.168800354004 ms
('current:', 7)
('i=0', ' j=-1', 'current=7', 'arr[i]=7', 'arr[j]=8', 'arr[]=', [7, 11, 9, 10, 12, 13, 8])
('current:', 11)
('i=1', ' j=0', 'current=11', 'arr[i]=11', 'arr[j]=7', 'arr[]=', [7, 11, 9, 10, 12, 13, 8])
('current:', 9)
('i=2', ' j=1', 'current=9', 'arr[i]=9', 'arr[j]=11', 'arr[]=', [7, 11, 9, 10, 12, 13, 8])
('current:', 10)
('i=3', ' j=2', 'current=10', 'arr[i]=10', 'arr[j]=9', 'arr[]=', [7, 9, 9, 10, 12, 13, 8])
('current:', 12)
('i=4', ' j=3', 'current=12', 'arr[i]=12', 'arr[j]=10', 'arr[]=', [7, 9, 9, 10, 12, 13, 8])
('current:', 13)
('i=5', ' j=4', 'current=13', 'arr[i]=13', 'arr[j]=12', 'arr[]=', [7, 9, 9, 10, 12, 13, 8])
('current:', 8)
('i=6', ' j=5', 'current=8', 'arr[i]=8', 'arr[j]=13', 'arr[]=', [7, 9, 9, 10, 12, 13, 8])
[7, 8, 9, 9, 10, 12, 8]
time:0.189781188965 ms
('i=1', ' j=0', 'arr[i]=11', 'new_arr[j]=7', 'new_arr[]=', [7])
('i=2', ' j=0', 'arr[i]=9', 'new_arr[j]=7', 'new_arr[]=', [7, 11])
('i=2', ' j=1', 'arr[i]=9', 'new_arr[j]=11', 'new_arr[]=', [7, 11])
('i=3', ' j=0', 'arr[i]=10', 'new_arr[j]=7', 'new_arr[]=', [7, 9, 11])
('i=3', ' j=1', 'arr[i]=10', 'new_arr[j]=9', 'new_arr[]=', [7, 9, 11])
('i=3', ' j=2', 'arr[i]=10', 'new_arr[j]=11', 'new_arr[]=', [7, 9, 11])
('i=4', ' j=0', 'arr[i]=12', 'new_arr[j]=7', 'new_arr[]=', [7, 9, 10, 11])
('i=4', ' j=1', 'arr[i]=12', 'new_arr[j]=9', 'new_arr[]=', [7, 9, 10, 11])
('i=4', ' j=2', 'arr[i]=12', 'new_arr[j]=10', 'new_arr[]=', [7, 9, 10, 11])
('i=4', ' j=3', 'arr[i]=12', 'new_arr[j]=11', 'new_arr[]=', [7, 9, 10, 11])
('i=5', ' j=0', 'arr[i]=13', 'new_arr[j]=7', 'new_arr[]=', [7, 9, 10, 11, 12])
('i=5', ' j=1', 'arr[i]=13', 'new_arr[j]=9', 'new_arr[]=', [7, 9, 10, 11, 12])
('i=5', ' j=2', 'arr[i]=13', 'new_arr[j]=10', 'new_arr[]=', [7, 9, 10, 11, 12])
('i=5', ' j=3', 'arr[i]=13', 'new_arr[j]=11', 'new_arr[]=', [7, 9, 10, 11, 12])
('i=5', ' j=4', 'arr[i]=13', 'new_arr[j]=12', 'new_arr[]=', [7, 9, 10, 11, 12])
('i=6', ' j=0', 'arr[i]=8', 'new_arr[j]=7', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('i=6', ' j=1', 'arr[i]=8', 'new_arr[j]=9', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('new_arr:', [7, 8, 9, 10, 11, 12, 13])
[7, 8, 9, 10, 11, 12, 13]
time:0.335931777954 ms
('i=1', ' new_len=0', 'arr[i]=11', 'new_arr[new_len]=7', 'new_arr[]=', [7])
('i=2', ' new_len=1', 'arr[i]=9', 'new_arr[new_len]=11', 'new_arr[]=', [7, 11])
('i=2', ' new_len=0', 'arr[i]=9', 'new_arr[new_len]=7', 'new_arr[]=', [7, 11])
('i=3', ' new_len=2', 'arr[i]=10', 'new_arr[new_len]=11', 'new_arr[]=', [7, 9, 11])
('i=3', ' new_len=1', 'arr[i]=10', 'new_arr[new_len]=9', 'new_arr[]=', [7, 9, 11])
('i=4', ' new_len=3', 'arr[i]=12', 'new_arr[new_len]=11', 'new_arr[]=', [7, 9, 10, 11])
('i=5', ' new_len=4', 'arr[i]=13', 'new_arr[new_len]=12', 'new_arr[]=', [7, 9, 10, 11, 12])
('i=6', ' new_len=5', 'arr[i]=8', 'new_arr[new_len]=13', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('i=6', ' new_len=4', 'arr[i]=8', 'new_arr[new_len]=12', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('i=6', ' new_len=3', 'arr[i]=8', 'new_arr[new_len]=11', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('i=6', ' new_len=2', 'arr[i]=8', 'new_arr[new_len]=10', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('i=6', ' new_len=1', 'arr[i]=8', 'new_arr[new_len]=9', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('i=6', ' new_len=0', 'arr[i]=8', 'new_arr[new_len]=7', 'new_arr[]=', [7, 9, 10, 11, 12, 13])
('new_arr:', [7, 8, 9, 10, 11, 12, 13])
[7, 8, 9, 10, 11, 12, 13]
time:0.219106674194 ms
"""